rfcs/000-search-revamp.md

```diff
@@ -0,0 +1,124 @@
+keep in mind that there's two kind of fields: single-valued fields (table name, schema name) and multi-valued fields (columns, tags, badges).
+Our proposed tradeoff between expressiveness and simplicity is to combine filter values for single-valued 
+fields with OR (e.g. "staging,production" filter applied to the "schema" field)
+whereas for multi-valued fields we'd give users an options to choose AND or OR 
+(a user would choose AND/OR to apply the filters "product_id,user_id" to the "column" field)
+
+#### Negation filtering
+Another highly requested feature is specifying negative filters - e.g. "schema is not temp, tmp, or temporary".
+As above, we want to avoid going down the rabbit hole of full-blown logical expressions. Likely, we'd implement
+this on the frontend as a separate set of filter boxes below the current set. Stay tuned for designs. 
+
+#### Type-ahead suggestions for filters
+Because values entered as filters are not interpreted or analyzed (this is the current behavior and is not expected to change),
+it can be hard for users to remember/guess the exact values to filter for. We'd like to provide typeahead
+suggestions that will suggest the most likely filter value given a prefix.
+
+
+## Backend Changes
+As part of this work, most of the elasticsearch [proxy](https://github.com/amundsen-io/amundsen/blob/main/search/search_service/proxy/elasticsearch.py)
+would be re-written, while keeping the API of the search service the same (in part to allow Atlas to stay unchanged).
+
+#### Upgrade Elasticsearch to v7
+Elasticsearch queries/mappings are generally written to a specific version, since a number of features are typically
+added and removed in each new release. We suggest moving to v7 as part of this revamp. How to gracefully handle 
+the backwards incompatibility for folks not ready to adopt a new ES version is an open question. 
+One option is to keep the existing ES proxy for v6 uses and add all our improvements in new code.
+
+#### Filtered vs unfiltered search unification
+Filtered search was added on top of regular search and was implemented separately via separate endpoints. 
+It also uses a different (and incorrect) elasticsearch [query_string](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-query-string-query.html) command. 
+We plan to unify the implementations, so both use the same top-level elasticsearch command either with or without an additional filter specifier. 
+
+#### Replace JSON in queries/mappings with Elasticsearch library objects
+The requests made to Elasticsearch are constructed with either hardcoded (in the case of mappings) or templated JSON. 
+This makes it more difficult to compose and understand requests, 
+but more importantly it makes it very difficult to extend OSS code. 
+The Elasticsearch-dsl library includes classes for constructing requests by putting together native python objects.
```

Comment:
I think one aspect that could be also raised is how we approach mapping definitions.

Now we have separate mappings for each type of object (table,feature,user,dashboard) and I think the refactor could also improve on that. We can have one mapping (or even better - template) that would span all the fields for all objects. It's a common pattern and it would make reusability of new analyzers/tokenizers/filters much easier.

Reply:
This is an excellent suggestion and my instinct is strongly in favor. Let me think on this and draw up the pros/cons and how to best implement.